autoprecompiles/src/low_degree_bus_interaction_optimizer.rs

```diff
@@ -0,0 +1,562 @@
+    /// individual inputs.
+    /// The outputs are generated by asking the bus interaction handler for each input assignment.
+    ///
+    /// If at any time (1) the inputs violate a constraint or (2) the outputs are not unique,
+    /// an error is yielded.
+    fn concrete_input_output_pairs<'b>(
+        &'b self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+        input_output_pair: &'b SymbolicInputOutputPair<T, V>,
+    ) -> impl Iterator<Item = Result<(Vec<T>, T), ()>> + 'b {
+        let bus_interaction = bus_interaction.to_range_constraints(self.solver);
+
+        // Consider all possible input assignments, which is the cross product of all allowed values.
+        let input_assignments = input_output_pair
+            .inputs
+            .iter()
+            .map(move |input| {
+                input
+                    .range_constraint
+                    .allowed_values()
+                    .map(|v| (input.index, v))
+                    .collect_vec()
+            })
+            .multi_cartesian_product();
+
+        // For each input assignment, try it and ask the bus interaction handler if there
+        // is a unique output assignment.
+        input_assignments.map(move |assignment| {
+            // Set all inputs to concrete values
+            let mut bus_interaction = bus_interaction.clone();
+            for (i, value) in assignment.iter() {
+                bus_interaction.payload[*i] = RangeConstraint::from_value(*value);
+            }
+
+            let inputs = assignment.into_iter().map(|(_i, value)| value).collect();
+
+            // Get the output from the bus interaction handler, if it exists and is unique.
+            let output = self
+                .bus_interaction_handler
+                .handle_bus_interaction_checked(bus_interaction)
+                // If the assignment violates a constraint, return an error.
+                .map_err(|_| ())?
+                .payload
+                .get(input_output_pair.output.index)
+                .unwrap()
+                .try_to_single_value()
+                // If the output is not unique, return an error.
+                .ok_or(())?;
+            Ok((inputs, output))
+        })
+    }
+}
+
+/// Represents a low-degree function, mapping a list of inputs to a single output.
+type LowDegreeFunction<T, V> = Box<dyn Fn(Vec<GroupedExpression<T, V>>) -> GroupedExpression<T, V>>;
+
+/// The maximum size of the input domain for low-degree functions.
+const MAX_DOMAIN_SIZE: u64 = 256;
+
+#[derive(Clone, Debug)]
+struct Replacement<T: FieldElement, V> {
+    polynomial_constraint: GroupedExpression<T, V>,
+    range_constraints: RangeConstraints<T, V>,
+}
+
+struct SymbolicInput<T: FieldElement, V> {
+    index: usize,
+    expression: GroupedExpression<T, V>,
+    range_constraint: RangeConstraint<T>,
+}
+
+struct SymbolicOutput<T: FieldElement, V> {
+    index: usize,
+    expression: GroupedExpression<T, V>,
```

Comment:
This is not the expression to compute the output but the expression the output is equal to, right? Can you document that, please?